docs/source/cuda/examples.rst

```diff
@@ -3,6 +3,479 @@
+            data[tid] = shr[tid]
+
+This kernel can be run as follows and the same result is obtained through
+summing the host data by traditional means.
+
+.. code-block:: python
+
+   kernel.forall(len(a))(a, len(a))
+   print(a[0]) # array(523776)
+   sum(np.arange(1024)) # 523776
+
+This algorithm can be greatly improved upon by redesigning the inner loop
+to use sequential memory accesses, or even furthermore using strategies that
+keep more threads active and working since in this example most threads quickly
+become idle.
+
+.. _cuda_sessionization:
+
+Dividing Click Data into Sessions
+=================================
+
+
+A common problem in business analytics is that of grouping the activity of users of an online platform into
+sessions, called "sessionization". The idea is that users generally traverse through a website and perform
+various actions (clicking something, filling out a form, etc) in discrete groups. Perhaps a customer spends
+some time shopping for an item in the morning and then again at night - often the business is interested in
+treating these periods as separate interactions with their service, and this creates the problem of 
+programmatically splitting up activity in some agreed upon way.
+
+Here we'll illustrate how to write a group of numba kernels to solve this problem. We'll start with data 
+containing two fields: let user_id represent a unique ID corresponding to an individual customer, and let 
+action_time be a time that some unknown action was taken on the service. Right now, we'll assume there's 
+only one type of action, so all there is to know is when it happened.
+
+Our goal will be to create a new column called session_id which contains a label corresponding to a unique 
+session. We'll define the boundary between sessions as when there has been at least one hour between clicks.
+
+
+.. code-block:: python
+
+   from numba import cuda
+   import numpy as np
+
+   # set the timeout to one hour
+   session_timeout = np.int64(np.timedelta64('3600', 's'))
+   
+Here is a solution using Numba:
+
+.. code-block:: python
+
+   @cuda.jit
+   def sessionize(
+      user_id, timestamp, results, size
+   ):
+      gid = cuda.grid(1)
+      
+      
+      ### DETERMINE SESSION BOUNDARIES
+      is_sess_boundary = 0
+      if 0 < gid < size:
+         if (
+               user_id[gid] != user_id[gid - 1]
+               or timestamp[gid] - timestamp[gid - 1] > session_timeout
+         ):
+               is_sess_boundary = 1
+         else:
+               is_sess_boundary = 0
+      else:
+         is_sess_boundary = 1
+         
+         
+      ### DETERMINE SESSION LABELS
+      # this thread marks the start of a session
+      if gid < size:
+         if is_sess_boundary:
+               results[gid] = gid
+               look_ahead = 1
+
+               # check elements 'forward' of this one 
+               # until a new session boundary is found
+               while results[gid + look_ahead] == 0:
```

Comment:
How is it guaranteed that if another session boundary is written by the next thread, it is correctly seen by the current thread?

Reply:
Ahh, you are right. The reason is these used to be two separate kernels that I fused into one, forgetting that the synchronization that comes from two distinct kernel launches is actually critical. The first used to write the boundaries and the second built out the IDs. Will look into alternatives.